Feature/tao 5148 test runner storage

krampstudio opened the pull request:
_No rush on this, it's an old story I had the time to finish. Please take the time to perform additional tests_

Add a new module called the **testStore** it provides a unified store for the test runner and the plugin. The goal is to store all the data of a test in one store, and to hide the store management to the implementation logic. 
 - New method `getTestStore`  on the test runner instance. Gives you an instance of the test store.
 - New method `getPluginStore`  on the test runner instance. It's a convenience method to get a store inside a plugin. The given store will be managed like a dedicated store but will save and retrieve the data from the unified store.
 - Test runner providers have to implement the method `loadTestStore` if they want to manipulate a store
 - The testStore supports running tests (legacy mode) and will fallback to the old way to store data if needed (please test also this point)
 - The testStore takes care of the _volatile_ stores. No need anymore to listen for the event `storechange` to clear the store. The only action required is to declare the store (or sub store) as _volatile_. 


**Requires**
- [ ] https://github.com/oat-sa/tao-core/pull/1621

**Companion PRs**
 - [ ] https://github.com/oat-sa/extension-tao-testqti/pull/1204
 - [ ] https://github.com/oat-sa/extension-tao-test-runner-plugins/pull/46
 - [ ] https://github.com/oat-sa/extension-tao-act/pull/895 (optional)

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+     * @returns {Promise<Boolean>} true means unified
+     */
+    var selectStoreMode = function selectStoreMode(testId, preselectedBackend){
+        return store
```
I think this method could get a better name. We are not selecting anything here, we are just checking whether we have legacy-named stores for the current test. A "boolean-type" name seems more appropriated. Something like `hasLegacyStores` for example.

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+                });
+            }, preselectedBackend)
+            .then(function(foundStores){
+                return _.isArray(foundStores) && foundStores.length === 0;
```
Boolean casting could be done here to avoid having to do it on each consumer.

krampstudio replied:
Not sure to get that

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+        var getMode = Promise.resolve();
+        if(_.isUndefined(unifiedStore)){
+            getMode = selectStoreMode(testId, preselectedBackend).then(function(result){
+                unifiedStore = !!result;
```
see previous comment

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+                    throw new TypeError('A store name must be provided to get the store');
+                }
+
+                if(!_.contains(storeNames, storeName)){
```
Readability could be improved by naming the first var `allStoreNames` (for ex) rather than having 2 vars with only 1 letter difference.

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+                if(!_.contains(storeNames, storeName)){
+                    storeNames.push(storeName);
+                }
+                return getMode.then(function(){
```
since `getMode` is not returning anything and has a side-effect (setting a 'global' var) then it could be renamed `setMode` (or changed to return the mode as a `resolve` parameter)

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+                         * Get all store items
+                         * @returns {Promise<Object>} with a collection of items
+                         */
+                        getItems : function getItems(){
```
in the same fashion as above, I prefer avoiding having two methods with really close names. `getAllItems()` would mark the difference, for ex, but the obvious advantage here is that it matches the store api so maybe it's better to keep it this way.

krampstudio replied:
The goal is to wrap the `store` API, so I'll keep as it is. And in any case, I'm not really agree since the plural is sufficient to mark the difference.

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+                         * @param {String} key - the item key
+                         * @returns {Promise<Boolean>} with true in resolve if removed
+                         */
+
```
extra line

no-chris commented on views/js/test/runner/testStore/test.html:
```diff
@@ -0,0 +1,34 @@
+<html>
+    <head>
+        <meta charset="utf-8">
+        <title>Test Runner - Navigator</title>
```
Wrong title

no-chris commented on views/js/runner/testStore.js:
```diff
@@ -0,0 +1,295 @@
+             * @param {String} storeId - the id to check
+             * @returns {Promise<Boolean>} true if cleared
+             */
+            clearVolatileOnStoreChange : function clearVolatileOnStoreChange(storeId){
```
Also when reading this method title I was expecting some kind of listener, or that it would be bound to a listener, but it seems that it is just called as a regular function in the provider init. To avoid confusion it may be better named something like `clearVolatileIfStoreChanged`

krampstudio commented:
I've added explicitly the store modes, to improve the readability. Changed the method names but I've still kept the store mode within the closure. I'm agree this can be viewed as a side effect within the scope, but this is scoped, not exposed and the inner function is just an artifact. The exposed method has no side effect and maintening a state prevent's you from reading the db multiple time.
